### PR TITLE
Examples grover

### PR DESCRIPTION
**Pull request by AdrianPerezSalinas**
Merging over `grovermodel`

### CODE REVIEW

**Comment by scarrazza** [comment]
@AdrianPerezSalinas thanks for this. I have noticed that the sphinx documentation is not rendering properly, could you please have a look?

File: examples/grover/README.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
+# A General Grover Model
+
+This examples provide information to run a general Grover model accesible as
```
Comment (scarrazza) [inline examples/grover/README.md:3]: These / or better "The examples presented here..."

File: examples/grover/README.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+- For Grover's original search algorithm: [arXiv:quant-ph/9605043](https://arxiv.org/abs/quant-ph/9605043)
+- For the iterative version with unknown solutions:[arXiv:quant-ph/9605034](https://arxiv.org/abs/quant-ph/9605034)
+- For the Grover algorithm with any superposition:[arXiv:quant-ph/9712011] (https://arxiv.org/abs/quant-ph/9712011)
+
+The arguments of the model are
+
+- oracle (`qibo.core.circuit.Circuit`): quantum circuit that flips
+    the sign using a Grover ancilla initialized with -X-H-. Grover ancilla
+    expected to be last qubit of oracle circuit.
```
Comment (scarrazza) [inline examples/grover/README.md:16]: the last?

File: examples/grover/README.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+- target_amplitude (float): absolute value of the amplitude of the target state. Only for
+    advanced use and known systems.
+- check (function): function that returns True if the solution has been
+    found. Required of iterative approach.
+    First argument should be the bitstring to check.
+- check_args (tuple): arguments needed for the check function.
+    The found bitstring not included.
+- iterative (bool): force the use of the iterative Grover
+
+We provide three different examples to run Grover with
```
Comment (scarrazza) [inline examples/grover/README.md:37]: with:

File: examples/grover/README.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+### Example 1: standard Hadamard superposition and simple oracle. 
+
+In this first example we show how to use a standard Grover search where the
+search space is an equally weighted superposition of quantum states and the oracle
+is simply defined through a layer of Hadamard gates. This example makes no use of 
+ancilla qubits, so the syntaxis simplifies greatly.
```
Comment (scarrazza) [inline examples/grover/README.md:44]: the syntax, or the command lines / function calls?

File: examples/grover/README.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+For the iterative case, the corresponding code is 
+```buildoutcfg
+grover = Grover(oracle, superposition_qubits=qubits, check=check, check_args=(num_1,))
+solution, iterations = grover()
+```
+
+
+### Example 3: Ancillas for superposition and oracle, setting size of search space
+
+In this third example we create a Grover model with two components
```
Comment (scarrazza) [inline examples/grover/README.md:93]: components:

File: examples/grover/example_1.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+
```
Comment (scarrazza) [inline examples/grover/example_1.py:1]: remove empty line

File: examples/grover/example_1.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+oracle.add(gates.X(5).controlled_by(*range(5)))
+# Create superoposition circuit. Ex: Full superposition over 5 qubits.
+
+# Generate and execute Grover class
+grover = Grover(oracle, superposition_circuit=superposition, number_solutions=1)
+
+solution, iterations = grover()
+
+print('The solution is', solution)
+print('Number of iterations needed:', iterations)
```
Comment (scarrazza) [inline examples/grover/example_1.py:21]: add empty line.

**Review by scarrazza — CHANGES_REQUESTED** [review]
Thanks for this implementation, I have included some comments above, in general however would be great if you could convert the file to PEP8 convetion and include these files in `test_examples.py`.

**Comment by AdrianPerezSalinas** [comment]
Hi @scarrazza , 

I think everything is fixed now, please let me know if there is anything wrong

File: examples/grover/1/main.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+    print('The solution is', solution)
+    print('Number of iterations needed:', iterations)
+
+    return solution, iterations
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--qubits", default=5, type=int)
```
Comment (scarrazza) [inline examples/grover/1/main.py:36]: @AdrianPerezSalinas this example is not working for me, it crashes at line 20.
Could you please rename `qubits` in `nqubits` in order to match the syntax of other examples?

File: examples/grover/README.md [hunk 9]
```diff
@@ -0,0 +1,123 @@
+# A General Grover Model
+
```
Comment (scarrazza) [inline examples/grover/README.md:2]: Please include a reference link to the code, as we do in all tutorials.

**Review by scarrazza — COMMENTED** [review]
Let me suggest to move the examples to:
- example1.py
- example2.py
- example3.py

File: examples/grover/1/main.py (hunk 8 above)
Comment (AdrianPerezSalinas) [reply examples/grover/1/main.py:36]: Does the problem come from this mismatch? I have tested it and it works

Comment (AdrianPerezSalinas) [reply examples/grover/1/main.py:36]: How is the `test_examples.py` file run? I cannot see it clearly and I think here is where the problem lies

File: examples/grover/README.md (hunk 9 above)
Comment (AdrianPerezSalinas) [reply examples/grover/README.md:3]: Done

**Pull request merged**